### ISSUE TITLE
File upload fails due to attempt to serialize MultipartFile

### ISSUE THREAD

**Issue by vasilich6107** [zino-app/graphql-flutter#723]
**Describe the issue**
File upload fails due to attempt to serialize MultipartFile

* Was there a **regression**?
no

**To Reproduce**
create a request with MultipartFile

**Expected behavior**
query executes without errors

### Other useful/optional fields

It happens here
https://github.com/zino-app/graphql-flutter/blob/modularization/packages/graphql/lib/src/core/query_manager.dart#L195

This could be fixed with setting `fetchPolicy: FetchPolicy.noCache,` for your mutation

**Comment by micimize** [comment]
@vasilich6107 ah, so it's trying to serialize and then deserialize the `MultipartFile`. Can you add a stacktrace? I wonder what the best behavior here would be.

**Comment by vasilich6107** [comment]
```
_JsonStringifier.writeObject (json.dart:687)
_JsonStringifier.writeMap (json.dart:768)
_JsonStringifier.writeJsonValue (json.dart:723)
_JsonStringifier.writeObject (json.dart:678)
_JsonStringifier.writeList (json.dart:735)
_JsonStringifier.writeJsonValue (json.dart:717)
_JsonStringifier.writeObject (json.dart:678)
_JsonStringifier.writeMap (json.dart:768)
_JsonStringifier.writeJsonValue (json.dart:723)
_JsonStringifier.writeObject (json.dart:678)
_JsonStringifier.writeMap (json.dart:768)
_JsonStringifier.writeJsonValue (json.dart:723)
_JsonStringifier.writeObject (json.dart:678)
_JsonStringStringifier.printOn (json.dart:876)
_JsonStringStringifier.stringify (json.dart:861)
JsonEncoder.convert (json.dart:261)
JsonCodec.encode (json.dart:171)
fieldNameWithArguments (field_name_with_arguments.dart:64)
denormalize.denormalizeNode.<anonymous closure> (denormalize.dart:101)
ListMixin.fold (list.dart:218)
denormalize.denormalizeNode (denormalize.dart:98)
denormalize (denormalize.dart:128)
NormalizingDataProxy.readQuery (_normalizing_data_proxy.dart:60)
QueryManager._resolveQueryOnNetwork (query_manager.dart:197)
<asynchronous gap>
QueryManager.fetchQueryAsMultiSourceResult (query_manager.dart:152)
ObservableQuery.fetchResults (observable_query.dart:182)
MutationState.runMutation (mutation.dart:107)
MutationState.runMutation (mutation.dart:1)
_Closure.call (function.dart:1)
_RhReportState._updateReport (rh_report.dart:153)
<asynchronous gap>
_RhReportState._buildFooter.<anonymous closure>.<anonymous closure>.<anonymous closure>.<anonymous closure> (rh_report.dart:305)
_Closure.call (function.dart:1)
_InkResponseState._handleTap (ink_well.dart:992)
_InkResponseState.build.<anonymous closure> (ink_well.dart:1098)
_Closure.call (function.dart:1)
GestureRecognizer.invokeCallback (recognizer.dart:184)
TapGestureRecognizer.handleTapUp (tap.dart:524)
BaseTapGestureRecognizer._checkUp (tap.dart:284)
BaseTapGestureRecognizer.handlePrimaryPointer (tap.dart:219)
PrimaryPointerGestureRecognizer.handleEvent (recognizer.dart:477)
PrimaryPointerGestureRecognizer.handleEvent (recognizer.dart:1)
_Closure.call (function.dart:1)
PointerRouter._dispatch (pointer_router.dart:78)
PointerRouter._dispatchEventToRoutes.<anonymous closure> (pointer_router.dart:124)
_LinkedHashMapMixin.forEach (compact_hash.dart:377)
PointerRouter._dispatchEventToRoutes (pointer_router.dart:122)
PointerRouter.route (pointer_router.dart:108)
GestureBinding.handleEvent (binding.dart:220)
GestureBinding.dispatchEvent (binding.dart:200)
GestureBinding._handlePointerEvent (binding.dart:158)
GestureBinding._flushPointerEventQueue (binding.dart:104)
GestureBinding._handlePointerDataPacket (binding.dart:88)
GestureBinding._handlePointerDataPacket (binding.dart:1)
_rootRunUnary (zone.dart:1206)
_rootRunUnary (zone.dart:1)
_CustomZone.runUnary (zone.dart:1100)
_CustomZone.runUnaryGuarded (zone.dart:1005)
_invoke1 (hooks.dart:267)
_dispatchPointerDataPacket (hooks.dart:176)
```

**Comment by aaahrens** [comment]
Thank you so much for providing the hotfix for this.

**Comment by micimize** [comment]
Just published `4.0.0-alpha.5` which should fix this. It adds a `sanitizeVariables` argument to `GraphQLCache`, which then `jsonDecode(jsonEncode(variables, toEncodable: sanitizeVariables))`.

The default sanitizer stringifies `filename + hashCode` so that the same request with different files won't be cached equivalently. I haven't worked at all with file uploads, so let me know if this is sensible:
```dart
/// Converts [MultipartFile]s to a string representation containing hashCode. Default argument to [variableSanitizer]
Object sanitizeFilesForCache(dynamic object) {
  if (object is MultipartFile) {
    return 'MultipartFile(filename=${object.filename} hashCode=${object.hashCode})';
  }
  return object.toJson();
}
```

This might also preemptively the issue @vasilich6107 was having with `v3` where he couldn't re-request file upload requests, which may have been due to us [using `filename` as part of the cache key](https://github.com/zino-app/graphql-flutter/blob/d17b57b52d3cb985a8c8d7025a242bbcddecabfb/packages/graphql/lib/src/core/raw_operation_data.dart#L77-L90)

**Comment by vasilich6107** [comment]
Just found time to test.
Now it works without errors.
